# omicsdata19
Materials for Omics Data Analysis - MESIO UPC 2019

- Where to look for anything


https://cran.r-project.org/ The comprehensive R archive network (R and R packages)

https://www.bioconductor.org/ Bioconductor (R packages for (mainly) bioinformatics and biostatistics data analysis)

https://bioconductor.org/help/ Bioconductor learning and support resources (EXTREMELY RECOMMENDED)

https://www.biostars.org/ Biostars knowledge base

http://seqanswers.com/ NGS knowledge base


- Illumina sequencing



https://www.youtube.com/watch?v=fCd6B5HRaZ8

https://www.youtube.com/watch?v=jFCD8Q6qSTM

https://s3.amazonaws.com/bitesizebio/content/uploads/2012/08/figure1.png



- Illumina RAW data and analysis



https://www.youtube.com/watch?v=l4BAfRekohk



- The FASTQ format



https://en.wikipedia.org/wiki/FASTA_format

https://en.wikipedia.org/wiki/FASTQ_format

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2847217/



- Short Read Aligners (Bowtie, Bowtie2, Tophat, Tophat2, BWA, Maq, STAR, SALMON, Hisat2â€¦)



Bowtie - DNA-Seq, ChIP-Seq, etc, end-to-end

http://bowtie-bio.sourceforge.net/index.shtml

http://bowtie-bio.sourceforge.net/bowtie2/index.shtml



Tophat - STAR - RNASeq - spliced (gapped) alignments

https://ccb.jhu.edu/software/tophat/index.shtml

https://github.com/alexdobin/STAR



- The SAM-BAM format



http://www.metagenomics.wiki/tools/samtools/bam-sam-file-format

https://bioinformatics.stackexchange.com/questions/14/what-is-the-difference-between-fasta-fastq-and-sam-file-formats



- FASTQC



https://www.bioinformatics.babraham.ac.uk/projects/fastqc/



- Cleaning FastQ files: sequencing 'artifacts'. cutadapt, trimmomatic



https://cutadapt.readthedocs.io/en/stable/guide.html

http://www.usadellab.org/cms/?page=trimmomatic



- Cleaning and filtering SAM-BAM files: samtools, sambamba, picard



http://www.htslib.org/

http://lomereiter.github.io/sambamba/

https://broadinstitute.github.io/picard/

