# Omics Data 19, May 8th 2019

# A) Continue with...

- Hands-on tutorials, RNASeq with DESeq2

  https://github.com/hbctraining/Intro-to-rnaseq-hpc-salmon/blob/master/lectures/RNAseq-analysis-methods.pdf (workflow diagram)

  https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
  (a good explanation on the nature of counts, overdispersion and the NB distribution)

  http://pedagogix-tagc.univ-mrs.fr/courses/ASG1/practicals/rnaseq_diff/rnaseq_diff.html
  (similar to the above)
  
  And
  
  http://folk.uio.no/jonbra/MBV-INF4410_2017/exercises/2017-12-07_R_DESeq2_exercises_with_results.html
  (a sort of 'complete' example from reading counts to obtaining a DE results table)

# B) Explore additional tutorial

- RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR
  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4937821/
  
- From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline  
  https://f1000research.com/articles/5-1438/v2
  
